File: Medium/1387. Sort Integers by The Power Value/1387. Sort Integers by The Power Value.js
/**
 * @param {number} lo
 * @param {number} hi
 * @param {number} k
 * @return {number}
 */
var getKth = function (lo, hi, k) {
    const nums = [];

    for (let i = lo; i <= hi; i++) {
        nums.push([i, calculatePower(i)]);
    }

    nums.sort((a, b) => {
        if (a[1] !== b[1]) {
            return a[1] - b[1];
        } else {
            return a[0] - b[0];
        }
    });

    return nums[k - 1][0];
};

const calculatePower = x => {
    let steps = 0;
    while (x !== 1) {
        if (x % 2 === 0) {
            x /= 2;
        } else {
            x = 3 * x + 1;
        }
        steps++;
    }
    return steps;
};

File: Medium/1387. Sort Integers by The Power Value/1387. Sort Integers by The Power Value.rs
impl Solution {
    pub fn get_kth(lo: i32, hi: i32, k: i32) -> i32 {
        fn calculate_power(x: i32) -> i32 {
            let mut steps = 0;
            let mut x = x;
            while x != 1 {
                if x % 2 == 0 {
                    x /= 2;
                } else {
                    x = 3 * x + 1;
                }
                steps += 1;
            }
            steps
        }

        let mut nums = Vec::new();

        for i in lo..=hi {
            nums.push((i, calculate_power(i)));
        }

        nums.sort_unstable_by(|a, b| a.1.cmp(&b.1).then(a.0.cmp(&b.0)));

        nums[k as usize - 1].0
    }
}
